crates/bevy_ecs/src/schedule_v3/schedule.rs

```diff
@@ -0,0 +1,1004 @@
+        schedule: Schedule,
+    ) -> Result<(), CheckInError> {
+        let label = label.dyn_clone();
+        match self.inner.get_mut(&label) {
+            Some(container) => match container.take() {
+                Some(_) => Err(CheckInError::AlreadyOccupied),
+                None => {
+                    *container = Some(schedule);
+                    Ok(())
+                }
+            },
+            None => Err(CheckInError::DoesNotExist),
+        }
+    }
+
+    /// Iterates all system change ticks and clamps any older than [`MAX_CHANGE_AGE`](crate::change_detection::MAX_CHANGE_AGE).
+    /// This prevents overflow and thus prevents false positives.
+    ///
+    /// **Note:** Does nothing if the [`World`] counter has not been incremented at least [`CHECK_TICK_THRESHOLD`](crate::change_detection::CHECK_TICK_THRESHOLD)
+    /// times since the previous pass.
+    // TODO: perf
+    #[doc(hidden)]
+    pub fn check_change_ticks(&mut self, change_tick: u32, last_check_tick: u32) {
+        #[cfg(feature = "trace")]
+        let _span = bevy_utils::tracing::info_span!("check ticks").entered();
+        for schedule in self.inner.values_mut().flatten() {
+            schedule.check_change_ticks(change_tick, last_check_tick);
+        }
+    }
+}
+
+/// A directed acyclic graph.
+struct DAG {
+    /// A directed acyclic graph.
+    graph: DiGraphMap<NodeId, ()>,
+    /// A cached topological ordering of the graph.
+    topsort: Vec<NodeId>,
+}
+
+impl Default for DAG {
+    fn default() -> Self {
+        Self {
+            graph: DiGraphMap::new(),
+            topsort: Vec::new(),
+        }
+    }
+}
+
+struct SystemSetMeta(BoxedSystemSet);
+
+impl SystemSetMeta {
+    pub fn new(set: BoxedSystemSet) -> Self {
+        Self(set)
+    }
+
+    pub fn name(&self) -> Cow<'static, str> {
+        format!("{:?}", &self.0).into()
+    }
+
+    pub fn is_system_type(&self) -> bool {
+        self.0.is_system_type()
+    }
+}
+
+struct ScheduleMeta {
+    system_set_ids: HashMap<BoxedSystemSet, NodeId>,
+    system_sets: HashMap<NodeId, SystemSetMeta>,
+    systems: HashMap<NodeId, BoxedSystem>,
+    conditions: HashMap<NodeId, Vec<BoxedCondition>>,
+
+    hierarchy: DAG,
+    dependency: DAG,
+    dependency_flattened: DAG,
+
+    default_set: Option<BoxedSystemSet>,
+    next_node_id: u64,
+    changed: bool,
+    uninit_nodes: Vec<(NodeId, DependencyInfo)>,
+}
+
+impl Default for ScheduleMeta {
+    fn default() -> Self {
+        Self {
+            system_set_ids: HashMap::new(),
+            system_sets: HashMap::new(),
+            systems: HashMap::new(),
+            conditions: HashMap::new(),
+            hierarchy: DAG::default(),
+            dependency: DAG::default(),
+            dependency_flattened: DAG::default(),
+            default_set: None,
+            next_node_id: 0,
+            changed: false,
+            uninit_nodes: Vec::new(),
+        }
+    }
+}
+
+impl ScheduleMeta {
+    pub fn new() -> Self {
```

Comment:
```suggestion
    pub const fn new() -> Self {
```

Reply:
Cannot make `const` because of `HashMap` and `DiGraphMap`.